base_vat_optional_vies/models/res_partner.py

```diff
@@ -0,0 +1,142 @@
+# License AGPL-3: Antiun Ingenieria S.L. - Antonio Espinosa
+# See README.rst file on addon root folder for more details
+
+import logging
+import re
+_logger = logging.getLogger(__name__)
+
+try:
+    import vatnumber
+except ImportError:
+    _logger.warning(
+        "VAT validation partially unavailable because the `vatnumber` Python "
+        "library cannot be found. Install it to support more countries, "
+        "for example with `easy_install vatnumber` or "
+        "`pip install vatnumber`.")
+    vatnumber = None
+
+from openerp import models, fields, api, _
+from openerp.exceptions import ValidationError
+
+
+class ResPartner(models.Model):
+    _inherit = 'res.partner'
+
+    vies_passed = fields.Boolean(
+        string="VIES validation passed", readonly=True)
+
+    def __init__(self, pool, cr):
+        super(ResPartner, self).__init__(pool, cr)
+        self._constraints = []
+
+    @api.constrains('vat')
+    def check_vat(self):
+        for partner in self:
+            if (not self.env.context.get('avoid_check_vat') and
+                    not partner.parent_id):
+                if not partner.validate_vat():
+                    raise ValidationError(_("VAT number is not valid"))
+
+    def button_check_vat(self, cr, uid, ids, context=None):
+        partner = self.browse(cr, uid, ids, context=context)
+        if not partner.validate_vat():
+            raise ValidationError(self._validation_error_msg())
+        return True
+
+    def _split_vat(self, vat, country=False):
+        """
+        @summary: Split Partner vat into country_code and number
+        @result: (vat_country, vat_number)
+        """
+        vat_country = False
+        vat_number = vat
+        if vat and re.match(r'[A-Za-z]{2}', vat):
+            vat_country = vat[:2].upper()
+            vat_number = vat[2:].replace(' ', '')
+        elif country:
+            vat_country = country
+        return vat_country, vat_number
+
+    def _validation_error_msg(self):
+        return _("The VAT number does not seem to be valid. \n"
+                 "Note: the expected format is 'CC##' "
+                 "(CC=Country Code, ##=VAT Number)")
+
+    @api.multi
+    def validate_vat(self):
+        self.ensure_one()
+        if self.company_id.vat_check_vies:
+            # VIES online check
+            check_func = self.vies_vat_optional_check
+        else:
+            # quick and partial off-line checksum validation
+            check_func = self.simple_vat_optional_check
+        vat_country, vat_number = self._split_vat(self.vat)
+        if not vat_country:
+            _logger.info("VAT country not found!")
+            raise ValidationError(self._validation_error_msg())
+        if not check_func(vat_country, vat_number):
+            _logger.info("VAT Number [%s] is not valid !" % vat_number)
+            return False
+        return True
+
+    @api.multi
+    def simple_vat_optional_check(self, country_code, vat_number):
```

Comment:
Yes, this is because Odoo original implementation fallbacks inside methods, instead of using methods that do just one thing, in order to be easier to inherit from them